src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs

```diff
@@ -280,21 +271,23 @@ public virtual string ReadString()
                 return string.Empty;
             }
 
-            _charBytes ??= new byte[MaxCharBytesSize];
+            _decoder ??= _encoding.GetDecoder();
             _charBuffer ??= new char[_maxCharsSize];
 
+            Span<byte> charBytes = stackalloc byte[MaxCharBytesSize];
+
             StringBuilder? sb = null;
             do
             {
                 readLength = ((stringLength - currPos) > MaxCharBytesSize) ? MaxCharBytesSize : (stringLength - currPos);
 
-                n = _stream.Read(_charBytes, 0, readLength);
+                n = _stream.Read(charBytes[..readLength]);
                 if (n == 0)
                 {
                     ThrowHelper.ThrowEndOfFileException();
                 }
 
-                charsRead = _decoder.GetChars(_charBytes, 0, n, _charBuffer, 0);
+                charsRead = _decoder.GetChars(charBytes[..n], _charBuffer, flush: false);
```

Comment:
The use of a stateful class like `Decoder` in this method does not seem right to me. Isn't each invocation of `ReadString` standalone?

Reply:
From the [docs]( https://learn.microsoft.com/en-us/dotnet/api/system.io.binaryreader.readstring?view=net-7.0):

> Reads a string from the current stream. The string is prefixed with the length, encoded as an integer seven bits at a time.

My understanding is that in this particular case the string must be prefixed with the length, so the job of the method is to read the bytes from the stream (and possibly modify it’s state like Position) and covert them into string.


The method does not read all bytes at a time (`readLength` can be different than `stringLength`) so the encoder should maintain it's state between the loop iterations:

https://github.com/dotnet/runtime/blob/be84df7642f21702b0e7859f0c134e4d6cfc67eb/src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs#L272

https://github.com/dotnet/runtime/blob/be84df7642f21702b0e7859f0c134e4d6cfc67eb/src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs#L289-L291

Now should it reset the state in the final iteration?

Checking for the leftovers and throwing would be considered a breaking change. Just resetting the state would be considered a breaking change as well, because someone in theory could rely on keeping this state.

Reply:
> Shouldn't we `Reset` it on each non-text operation?

I mean, if we call `Read()`, it will return the next character, and if we call `ReadUInt32(); Read()` after that, the decoder's state will be out-of-sync from the stream.

Reply:
> I mean, if we call Read(), it will return the next character, and if we call ReadUInt32(); Read() after that, the decoder's state will be out-of-sync from the stream.

Could you please add a unit test that exhibits the described behavior?

Reply:
I can't find an easy way to test it. Can we skip it for this PR?